tools/audio/sampleconv/src/codec/vadpcm.c

```diff
@@ -0,0 +1,1311 @@
+
+                if (nguess[i] != guess[i])
+                    changed = true;
+            }
+
+            if (!changed)
+                break;
+
+            int64_t score = scored_encode(nguess, prev_state, coef_tbl, order, npredictors, wanted_predictor,
+                                          wanted_scale, wanted_ix, frame_size);
+            if (score < bestScore) {
+                bestScore = score;
+                memcpy(bestGuess, nguess, sizeof(nguess));
+            }
+            maxMove *= 0.7;
+        }
+
+        if (bestScore == curScore)
+            return false; // No improvements along that line, give up.
+
+        curScore = bestScore;
+        memcpy(guess, bestGuess, sizeof(bestGuess));
+    }
+}
+
+static void
+get_bounds(int32_t *in, int32_t *prescaled, int32_t scale, int32_t *min_vals, int32_t *max_vals, int frame_size)
+{
+    int32_t minv, maxv;
+
+    if (frame_size == 9) {
+        minv = -8;
+        maxv = 8 - 1;
+    } else {
+        minv = -2;
+        maxv = 2 - 1;
+    }
+
+    for (int32_t i = 0; i < 16; i++) {
+        int32_t lo = in[i] - scale / 2 - scale;
+        int32_t hi = in[i] + scale / 2 + scale;
+
+        if (prescaled[i] == minv)
+            lo -= scale;
+        else if (prescaled[i] == maxv)
+            hi += scale;
+
+        min_vals[i] = lo;
+        max_vals[i] = hi;
+    }
+}
+
+#define VADPCM_BRUTEFORCE_LIMIT 100000
+
+static int32_t
+bruteforce(int32_t guess[16], uint8_t input[9], int32_t decoded[16], int32_t prescaled[16], int32_t prev_state[16],
+           int32_t ***coef_tbl, int32_t order, int32_t npredictors, int frame_size)
+{
+    int32_t scale = (input[0] >> 4) & 0xF;
+    int32_t predictor = (input[0] >> 0) & 0xF;
+
+    int32_t min_vals[16];
+    int32_t max_vals[16];
+    get_bounds(decoded, prescaled, 1 << scale, min_vals, max_vals, frame_size);
+
+    int i = 0;
+    while (true) {
+        int64_t bestScore = -1;
+        int32_t bestGuess[16];
+
+        for (int tries = 0; tries < 100; tries++) {
+            permute(guess, decoded, prescaled, 1 << scale, frame_size);
+
+            int64_t score =
+                scored_encode(guess, prev_state, coef_tbl, order, npredictors, predictor, scale, prescaled, frame_size);
+            if (score == 0)
+                return true;
+
+            if (bestScore == -1 || score < bestScore) {
+                bestScore = score;
+                memcpy(bestGuess, guess, sizeof(bestGuess));
+            }
+        }
+
+        memcpy(guess, bestGuess, sizeof(bestGuess));
+
+        if (descent(guess, min_vals, max_vals, prev_state, coef_tbl, order, npredictors, predictor, scale, prescaled,
+                    frame_size))
+            return true;
+
+        i++;
+        if (i == VADPCM_BRUTEFORCE_LIMIT)
+            error("Bruteforce decoding failed");
+    }
+}
+
+static void
+vencodeframe(uint8_t *out_buf, int16_t *in_buf, int32_t *state, int32_t ***coef_tbl, int32_t order, int32_t npredictors,
```

Comment:
`my_encodeframe`, `scored_encode`, and `vencodeframe` all look pretty similar. In particular `my_encodeframe` and `vencodeframe` seem basically identical. If there's some subtle difference, maybe it's worth commenting about that?